File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.py
def solution(msg):
    answer = []
    dict_zip = {}
    
    for i in range(26):
        dict_zip[chr(65 + i)] = i+1
    number = 27
    
    # while msg 이렇게 하고 msg.pop(0) 이렇게 하려고 했는데 문자열은 pop을 못써서 idx로
    idx = 0
    while idx < len(msg):
        
        word = msg[idx]
        idx += 1
        
        while idx < len(msg) and (word + msg[idx]) in dict_zip:
            word += msg[idx]
            idx += 1
        
        if idx < len(msg):
            dict_zip[word + msg[idx]] = number
            number += 1
        answer.append(dict_zip[word])
        
    return answer

File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.java
import java.util.*;

class Solution {
    public List<Integer> solution(String msg) {
        Map<String, Integer> alphabet = new HashMap<>();
        int alpha = 1;
        for (char c = 'A'; c <= 'Z'; c++) {
            alphabet.put(String.valueOf(c), alpha++);
        }
        
        List<Integer> answer = new ArrayList<>();
        int idx = 0;
        while (idx < msg.length()) {
            String w = "";
            int i = idx;
            while (i < msg.length()) {
                String next_w = msg.substring(idx, i+1);
                if (alphabet.containsKey(next_w)) {
                    w = next_w;
                    i++;
                } else
                    break;
            }
            answer.add(alphabet.get(w));
            if (i < msg.length()) {
                String wc = msg.substring(idx, i+1);
                alphabet.put(wc, alpha++);
            }
            idx += w.length();
        }
        
        return answer;
    }
}